File: 13.SlidingWindow滑动窗口/leetcode_2904_shortestBeautifulSubstring.py
class Solution:
    def shortestBeautifulSubstring(self, s: str, k: int) -> str:
        pos = [i for i in range(len(s)) if s[i]=='1']
        cur = []
        res = '1'*(len(s)+1)
        for i in range(len(pos)):
            if len(cur) > 0:
                cur.append('0'*(pos[i]-pos[i-1]-1))
            cur.append('1')
            if i < k-1:
                continue
            curS = ''.join(cur)
            if len(curS) < len(res) or (len(curS)==len(res) and curS<res):
                res = curS
            cur.pop(0)
            if len(cur):
                cur.pop(0)
        return '' if res=='1'*(len(s)+1) else res

if __name__ == "__main__":
    sl = Solution()
    s = "100011001"
    k = 3
    print(sl.shortestBeautifulSubstring(s, k))

File: 13.SlidingWindow滑动窗口/leetcode_2904_shortestBeautifulSubstring.cpp
#include <iostream>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    string shortestBeautifulSubstring(string s, int k) {
        vector<int> pos;
        for (int i=0; i<s.size(); i++){
            if (s[i]=='1'){
                pos.push_back(i);
            }
        }

        string res = "";
        vector<string> cur;
        for (int i=0; i<pos.size(); i++){
            if (cur.size() > 0){
                cur.push_back(string(pos[i]-pos[i-1]-1, '0'));
            }
            cur.push_back("1");
            if (i < k-1){
                continue;
            }
            string curS = "";
            for (auto& c : cur){
                curS += c;
            }
            if (res == "" || curS.size()<res.size() || (curS.size()==res.size() && curS<res)){
                res = curS;
            }
            cur.erase(cur.begin());
            if (cur.size()){
                cur.erase(cur.begin());
            }
        }
        return res;
    }
};

int main(){
    Solution sl;
    string s = "10001101";
    int k = 3;
    cout << sl.shortestBeautifulSubstring(s, k) << endl;
}
